spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/services/impl/DefaultTaskServiceTests.java

```diff
@@ -317,11 +326,83 @@ public void verifyDataFlowUriProperty() throws Exception {
 		assertTrue(!appDeploymentProperties.containsKey("dataflowServerUri"));
 		assertTrue(!appDeploymentProperties.containsKey("DATAFLOW-SERVER-URI"));
 	}
+	@Test
+	@DirtiesContext
+	public void createFailedComposedTask() {
+		String dsl = "AAA && BBB";
+		initializeFailRegistry();
+		boolean isExceptionThrown = false;
+		try{
+			taskService.saveTaskDefinition("splitTask", dsl);
+		}
+		catch (IllegalArgumentException iae) {
+			isExceptionThrown = true;
+		}
+		assertTrue("IllegalArgumentException was expected to be thrown", isExceptionThrown);
+		assertFalse(wasTaskDefinitionCreated("splitTask"));
+		assertFalse(wasTaskDefinitionCreated("splitTask-AAA"));
+		assertFalse(wasTaskDefinitionCreated("splitTask-BBB"));
+	}
+
+	@Test
+	@DirtiesContext
+	public void createDuplicateComposedTask() {
+		String dsl = "AAA && BBB";
+		initializeSuccessfulRegistry();
+		boolean isExceptionThrown = false;
+		taskService.saveTaskDefinition("splitTask", dsl);
+		try{
+			taskService.saveTaskDefinition("splitTask", dsl);
+		}
+		catch (DuplicateTaskException de) {
+			isExceptionThrown = true;
+		}
+		assertTrue("DuplicateTaskException was expected to be thrown", isExceptionThrown);
+		assertTrue(wasTaskDefinitionCreated("splitTask"));
+		assertTrue(wasTaskDefinitionCreated("splitTask-AAA"));
+		assertTrue(wasTaskDefinitionCreated("splitTask-BBB"));
+	}
+
+	@Test
+	@DirtiesContext
+	public void createDuplicateChildTaskComposedTask() {
+		String dsl = "AAA && BBB";
+		initializeSuccessfulRegistry();
+		boolean isExceptionThrown = false;
+		taskService.saveTaskDefinition("splitTask-AAA", "AAA");
+		try{
+			taskService.saveTaskDefinition("splitTask", dsl);
+		}
+		catch (DuplicateTaskException de) {
+			isExceptionThrown = true;
+		}
+		assertTrue("DuplicateTaskException was expected to be thrown", isExceptionThrown);
+		assertFalse(wasTaskDefinitionCreated("splitTask"));
+		assertTrue(wasTaskDefinitionCreated("splitTask-AAA"));
+		assertFalse(wasTaskDefinitionCreated("splitTask-BBB"));
+	}
 
 	private void verifyTaskExistsInRepo(String taskName, String dsl) {
 		TaskDefinition taskDefinition = taskDefinitionRepository.findOne(taskName);
 
 		assertThat(taskDefinition.getName(), is(equalTo(taskName)));
 		assertThat(taskDefinition.getDslText(), is(equalTo(dsl)));
 	}
+
+	private void initializeSuccessfulRegistry() {
+		when(this.appRegistry.find(anyString(), any(ApplicationType.class)))
+				.thenReturn(new AppRegistration("some-name", task, URI.create("http://helloworld"), resourceLoader));
+	}
+
+	private void initializeFailRegistry() throws IllegalArgumentException{
```

Comment:
This fails on the first insert, doesn't it?  Which would mean that it's not really testing the rollback (since no inserts were successful in the first place).